### PR TITLE
Feature/distributed index 1289 gocb timeouts

### PR DESCRIPTION
**Pull request by tleyden**
Fixes:
- https://github.com/couchbase/sync_gateway/issues/1289
- https://github.com/couchbase/sync_gateway/issues/1247

This PR supercedes https://github.com/couchbase/sync_gateway/pull/1325

### CODE REVIEW

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 1]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+func shouldRetryGoCBOp(err error) bool {
+
+	if err == nil {
+		// no error, no need to retry
+		return false
+	}
+
+	return isRecoverableGoCBError(err)
+
+}
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go:453]: Is there value in having separate methods for shouldRetryGoCBOp and isRecoverableGoCBError?

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 2]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
-			// NOTE: some of the errors are misleading:
-			// https://issues.couchbase.com/browse/GOCBC-64
-			if getOp.Err == nil {
-				byteValue, ok := getOp.Value.(*[]byte)
-				if ok {
-					result[getOp.Key] = *byteValue
-				}
+		} else {
+			// if it's a recoverable error, then throw it in retry collection.
+			if isRecoverableGoCBError(getOp.Err) {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go:378]: Related to the question below - why isRecoverableGoCBError here, and not shouldRetryGoCBOp?

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 3]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+			}
+		}
+		batches[idxBatch] = batch
+	}
+
+	return batches
+
+}
+
+func isNotFoundError(err error) bool {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: Is this method obsolete now?

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 4]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+		for _, key := range keys {
+			batch = append(batch, key)
+			keys = keys[1:]
+			if len(batch) == batchSize {
+				// this batch is full
+				break
+			}
+		}
+		batches[idxBatch] = batch
+	}
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: It took me a few reads to understand how this loop worked.  It seems counter-intuitive to have multiple loops over the key set (one per batch), and modifying the keys slice while you iterate over it was also confusing the first time I read it.

It feels like it would be clearer code to just iterate over the keys array once as the outer loop, and switch batches as they fill up, but let me know if you've got a reason to prefer this approach.

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 5]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+}
+
+func createBatchesKeys(batchSize int, keys []string) [][]string {
+
+	// boundary checking
+	if len(keys) == 0 {
+		Warn("createBatchesKeys called with empty keys")
+		return [][]string{}
+	}
+	if batchSize == 0 {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go:428]: I guess <=0, since you're passing in an int.

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 6]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+		// return true to signal that this function needs to be retried
+		return true, nil, nil
+
+	}
+
+	return worker
+
+}
+
+func (bucket CouchbaseBucketGoCB) processBatch(keys []string, resultAccumulator map[string][]byte, retryKeys []string) error {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: This is specifically "processGetBatch", right?

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 7]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
-	if err != nil {
-		return nil, 0, err
 	}
-	return returnVal.([]byte), uint64(casGoCB), err
+
+	return worker
+
+}
+
+func (bucket CouchbaseBucketGoCB) processEntriesBatch(pendingKeys []string, entries []*sgbucket.BulkSetEntry, keyToIndex map[string]int) (error, []string) {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: Maybe rename to make it clear that this is processing a batch of set operations, compared to processBatch below?

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 8]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+	// Kick off retry loop
+	description := fmt.Sprintf("SetBulk with %v entries", len(entries))
+	err, _ = RetryLoop(description, worker, sleeper)
+
+	return err
+
+}
+
+func (bucket CouchbaseBucketGoCB) newSetBulkRetryWorker(entries []*sgbucket.BulkSetEntry) RetryWorker {
+
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go:176]: It feels like we're doing a lot of iterations over the entry set to prepare the actual bulk op requests:
- once to build pendingKeys
- once to build entriesKeysToIndex
- over pendingKeys, to build pendingKeysBatches
- in processEntriesBatch, again (per batch) to build the set of ops

Although separating out the iterations by functionality makes the code easier to read, this feels too much redundant processing work to me.

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 9]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+	pendingKeys := createPendingKeysFromEntries(entries)
+	entriesKeysToIndex := createEntriesKeyToIndex(entries)
+
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: I'm wondering whether we could just use entries as-is, instead of splitting into a keySet and an index lookup array.  Any reason that pendingKeysBatches and processEntriesBatch couldn't just work directly against entries (or a new slice that points at the underlying data for entries, if you need to do slice manipulation?)

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 10]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+		return err, retryKeys
+	}
+
+	for index, item := range items {
+		entry := entries[index]
+		switch item := item.(type) {
+		case *gocb.InsertOp:
+			entry.Cas = uint64(item.Cas)
+			entry.Error = item.Err
+			if item.Err != nil && !isCasFailure(item.Err) {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: Shouldn't be only be retrying on overflow/timeout here?

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 11]
```diff
@@ -93,91 +97,416 @@ func (bucket CouchbaseBucketGoCB) GetName() string {
...
+				retryKeys = append(retryKeys, entry.Key)
+			}
+		case *gocb.ReplaceOp:
+			entry.Cas = uint64(item.Cas)
+			entry.Error = item.Err
+			if item.Err != nil && !isCasFailure(item.Err) {
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: Same as above - only retry on isRecoverableGoCBError?

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go [hunk 12]
```diff
@@ -358,23 +745,51 @@ func (bucket CouchbaseBucketGoCB) Incr(k string, amt, def uint64, exp int) (uint
...
+
+	// If the retry loop returned a nil result, set to 0 to prevent type assertion on nil error
+	if result == nil {
+		result = uint64(0)
+	}
+
+	// Type assertion of result
+	cas, ok := result.(uint64)
+	if !ok {
+		LogPanic("Error doing type assertion of %v into a uint64", result)
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: Should we panic here, as opposed to just returning error?

File: src/github.com/couchbase/sync_gateway/db/kv_channel_storage.go [hunk 13]
```diff
@@ -204,8 +242,30 @@ func (b *BitFlagStorage) writeSingleBlockWithCas(entries []*LogEntry) error {
...
+			return nil, nil, err
+		}
+	}
+	localValue, err := block.Marshal()
+	if err != nil {
+		base.Warn("Unable to marshal channel block - cancelling block update")
+		return nil, block, errors.New("Error marshalling channel block")
+	}
+	return localValue, block, nil
+
```
Comment (adamcfraser) [inline src/github.com/couchbase/sync_gateway/db/kv_channel_storage.go:268]: Is this pulling in an unrelated commit?

Comment (tleyden) [inline src/github.com/couchbase/sync_gateway/db/kv_channel_storage.go:268]: It's part of this commit: https://github.com/couchbase/sync_gateway/commit/f416d4442accebceff8d713e88d087050c66a99c

which got lopped into this PR due to code / manual testing dependencies.

File: src/github.com/couchbase/sync_gateway/base/bucket_gocb.go (hunk 9 above)
Comment (tleyden) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: Agreed there is a lot of redundant processing.  `pendingKeysBatches` could easily be reworked to work directly against entries, at the small expense of complicating the unit test a bit.  Reworking `processEntriesBatch` feels like more work, and I feel like the right way is to enhanc the `sg-bucket` `BulkSetEntry` struct, which I think would allow this code to be cleaner.  (but I think that definitely needs to be in a separate PR)

Comment (tleyden) [inline src/github.com/couchbase/sync_gateway/base/bucket_gocb.go]: Looking at this closer, I think the change is feasible w/o any changes to `sg-bucket.BulkSetEntry`.  I'm taking a pass at it.

**Comment by tleyden** [comment]
Functional test (currently queued): http://qa.sc.couchbase.com/view/mobile/job/centos-syncgateway-functional-tests-distributed-index/174/

**Comment by tleyden** [comment]
@adamcfraser this should be ready to merge:
- Commented out the boolean check related to tests
- Squashed into a single commit
